merino/middleware/logging.py

```
@@ -17,53 +21,67 @@
 PATTERN: Pattern = re.compile(r"/api/v[1-9]\d*/suggest$")
 
 
-class LoggingMiddleware(BaseHTTPMiddleware):
-    """Logging middleware for MozLog."""
+class LoggingMiddleware:
+    """An ASGI middleware for logging."""
+
+    def __init__(self, app: ASGIApp) -> None:
+        """Initilize."""
+        self.app = app
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send):
+        """Log requests."""
+        if scope["type"] != "http":  # pragma: no cover
+            await self.app(scope, receive, send)
+            return
+
+        async def send_wrapper(message: Message) -> None:
+            if message["type"] == "http.response.start":
+                request = Request(scope=scope)
+                if PATTERN.match(request.url.path):
+                    location = ctxvar_geolocation.get()
+                    ua = ctxvar_user_agent.get()
```

I'm pretty sure `ContextVar.get` returns an `Optional[T]` we should probably specify default values or have a guard to check against a None value

---

Yep, initially, I did provide a default value here just in case the get call fails. But after the second thought, I decided to take it out as the default value has already been set in the middleware, if the get ever fails, that should be a code error and it should fail fast and loud (as each request/task should always have a ctxvar associated as enforced by the middleware). Providing a default value here would otherwise silence that error inadvertently, which is probably not ideal in this case.

Does that make sense?

---

Yeah I was thinking for the purposes of static analysis not yelling at us. if we want some error reporting for a None result, then what about a guard for the results of the get and log an error if the are None?

---

Yep, but I was thinking about letting Sentry handle this.

Per its [doc](https://docs.python.org/3/library/contextvars.html#contextvars.ContextVar.get), `ContextVar.get()` will raise a `LookupError` if the value is not set and the default is not provided. And that exception (assuming it's rare) would be captured automatically by Sentry. For this particular request, an HTTP 500 error will be responded and it won't affect other requests since raising an exception inside of a middleware won't crash the app.

Do you think that's sufficient?